File: 2014.longest-subsequence-repeated-k-times.cs
/*
 * @lc app=leetcode id=2014 lang=csharp
 *
 * [2014] Longest Subsequence Repeated k Times
 */

// @lc code=start
using System;
using System.Collections.Generic;

public class Solution {
    private Dictionary<Tuple<int, string>, HashSet<string>> dp;
    /*
    Check if s is a subsequence of t
    */
    private bool IsSubsequence(string s, string t) {
        int p1 = 0, p2 = 0;
        while (p1 < s.Length && p2 < t.Length) {
            if (s[p1] == t[p2]) {
                p1++;
            }
            p2++;
        }
        return p1 == s.Length;
    }

    /*
    Given remaining available characters and number of remaining characters in possible subsequences, generate all possible subsequences
    */
    private HashSet<string> GenerateCandidates(int remainCharacters, string availableCharacters) {
        Tuple<int, string> t = new Tuple<int, string>(remainCharacters, availableCharacters);
        if (!dp.ContainsKey(t))
        {
            var result = new HashSet<string>(remainCharacters == 0 ? new[] { "" } : new HashSet<string>());
            if (remainCharacters > 0) {
                for (int i = 0; i < availableCharacters.Length; i++) {
                    if (i > 0 && availableCharacters[i] == availableCharacters[i - 1]) continue;
                    var furtherSearchSet = GenerateCandidates(remainCharacters - 1, availableCharacters.Substring(0, i) + availableCharacters.Substring(i + 1));
                    foreach (var s in furtherSearchSet) {
                        result.Add(availableCharacters[i] + s);
                    }
                }
            }
            dp.Add(t, result);            
        }
        return dp[t];
    }

    public string LongestSubsequenceRepeatedK(string s, int k) {
        var occur = new int[26];
        int n = s.Length;

        foreach (var c in s) {
            occur[c - 'a']++;
        }

        var startingString = "";
        for (int i = 25; i >= 0; i--) {
            startingString += new string((char)(i + 'a'), occur[i] / k);
        }

        dp = new Dictionary<Tuple<int, string>, HashSet<string>>();

        for (int i = Math.Min(n / k, startingString.Length); i > 0; i--) {
            var candidates = new List<string>(GenerateCandidates(i, startingString));
            candidates.Sort((a, b) => string.Compare(b, a, StringComparison.Ordinal));
            
            foreach (var cand in candidates) {
                string subseq = "";
                for (int j = 0; j < k; ++j) subseq += cand;
                if (IsSubsequence(subseq , s)) {
                    return cand;
                }
            }
        }

        return "";
    }
}

// @lc code=end

File: 2014.longest-subsequence-repeated-k-times.cpp
/*
 * @lc app=leetcode id=2014 lang=cpp
 *
 * [2014] Longest Subsequence Repeated k Times
 */

// @lc code=start
#include <vector>
#include <map>
#include <unordered_set>
#include <string>
#include <algorithm>

using namespace std;

class Solution {
private:
    map<pair<int, string>, unordered_set<string>> dp;

    bool IsSubsequence(const string& s, const string& t) {
        int p1 = 0, p2 = 0;
        while (p1 < s.length() && p2 < t.length()) {
            if (s[p1] == t[p2]) {
                p1++;
            }
            p2++;
        }
        return p1 == s.length();
    }

    unordered_set<string> GenerateCandidates(int remainCharacters, const string& availableCharacters) {
        pair<int, string> t = { remainCharacters, availableCharacters };
        if (dp.find(t) == dp.end()) {
            unordered_set<string> result;
            if (remainCharacters > 0) {
                for (int i = 0; i < availableCharacters.length(); i++) {
                    if (i > 0 && availableCharacters[i] == availableCharacters[i - 1]) continue;
                    auto furtherSearchSet = GenerateCandidates(remainCharacters - 1, availableCharacters.substr(0, i) + availableCharacters.substr(i + 1));
                    for (const auto& s : furtherSearchSet) {
                        result.insert(availableCharacters[i] + s);
                    }
                }
            }
            else
                result.insert(string {""});
            dp[t] = result;
        }
        return dp[t];
    }

public:
    string longestSubsequenceRepeatedK(const string& s, int k) {
        vector<int> occur(26, 0);
        int n = s.length();

        for (char c : s) {
            occur[c - 'a']++;
        }

        string startingString = "";
        for (int i = 25; i >= 0; i--) {
            startingString += string(occur[i] / k, static_cast<char>(i + 'a'));
        }

        dp.clear();

        for (int i = min(n / k, static_cast<int>(startingString.length())); i > 0; i--) {
            unordered_set<string> candidatesSet {GenerateCandidates(i, startingString)};
            vector<string> candidates (candidatesSet.begin(), candidatesSet.end());
            sort(candidates.begin(), candidates.end(), greater<string>());

            for (const auto& cand : candidates) {
                string subseq = "";
                for (int j = 0; j < k; ++j) subseq += cand;
                if (IsSubsequence(subseq, s)) {
                    return cand;
                }
            }
        }

        return "";
    }
};


// @lc code=end

